src/main/java/ArrayStack.java

```diff
@@ -88,4 +90,23 @@ public String toString() {
         builder.append("<-- Top");
         return builder.toString();
     }
+
+    @Override
+    public final boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ArrayStack<?> that = (ArrayStack<?>) o;
+        return top == that.top
+                && Arrays.equals(this.stack, 0, this.top, that.stack, 0, that.top);
+    }
+
+    @Override
+    public final int hashCode() {
+        int result = Objects.hash(top);
+        result = 31 * result;
+        for (int i = 0; i < Math.min(this.top, this.stack.length); i++){
+            result += stack[i].hashCode();
```

Comment:
Still not a great hashcode. Simply adding values leads to high collision rates. It's not a big deal for this I guess, but like. Ya know.